java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverridenWithCompareToCheck.java

```diff
@@ -56,35 +63,58 @@
   public void visitNode(Tree tree) {
     ClassTree classTree = (ClassTree) tree;
     if (isComparable(classTree)) {
-      boolean hasEquals = false;
       MethodTree compare = null;
 
       for (Tree member : classTree.members()) {
         if (member.is(Tree.Kind.METHOD)) {
           MethodTree method = (MethodTree) member;
-
-          if (isEqualsMethod(method)) {
-            hasEquals = true;
-          } else if (isCompareToMethod(method)) {
+          if (isCompareToMethod(method)) {
             compare = method;
           }
         }
       }
 
-      if (compare != null && !hasEquals) {
+      if (compare != null && !generatesEquals(classTree) && !implementsEquals(classTree)) {
         reportIssue(compare.simpleName(), "Override \"equals(Object obj)\" to comply with the contract of the \"compareTo(T o)\" method.");
       }
     }
   }
 
+  private static boolean implementsEquals(ClassTree classTree) {
+      return hasNotFinalEqualsMethod(classTree.symbol());
```

Comment:
As you removed the part where we look into the hierarchy of the current class (originally needed in `EqualsNotOverridenInSubclassCheck`), I think it's simpler to stay with the initial approach based on the `hasEquals` test.